### ISSUE TITLE
Yarn should use repository default branch instead of defaulting to master

### ISSUE THREAD

**Issue by ndrewtl** [yarnpkg/berry#3307]
### Self-service

- [X] I'd be willing to implement a fix

### Describe the bug

When resolving a package using the `github:` protocol, for instance `yarn add github:<orgname>/<reponame>`, yarn defaults to fetching the branch `master` instead of the default branch.

As per https://github.com/github/renaming, Github has been gradually transition from `master` to `main` as the default branch name for its repositories. As per https://docs.npmjs.com/cli/v7/commands/npm-install#description, when adding a package from a repository, if a branch or commit is not specified, NPM uses the default branch to fetch from. However, I recently renamed the default branch of one of my repositories from `master` to `main`, and yarn kept looking for branch `master`. It should use the default branch instead of hardcoding `master`. Here was my output:

```sh
_ YN0000: ┌ Project validation
_ YN0057: │ reponame: Resolutions field will be ignored
_ YN0000: └ Completed
_ YN0000: ┌ Resolution step
_ YN0001: │ Error: flexsearch@skiff-org/flexsearch: Unknown head ("master")
    at s (/path-to-repo/.yarn/releases/yarn-berry.js:2:138265)
    at v (/path-to-repo/.yarn/releases/yarn-berry.js:2:139315)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async F.resolvers.getCandidates (/path-to-repo/.yarn/releases/yarn-berry.js:2:141723)
    at async n.getCandidates (/path-to-repo/.yarn/releases/yarn-berry.js:2:337358)
    at async n.getCandidates (/path-to-repo/.yarn/releases/yarn-berry.js:2:337358)
    at async /path-to-repo/.yarn/releases/yarn-berry.js:2:359014
    at async Module.w (/path-to-repo/.yarn/releases/yarn-berry.js:2:418487)
    at async C (/path-to-repo/.yarn/releases/yarn-berry.js:2:358977)
    at async Promise.all (index 12)
_ YN0000: └ Completed in 0s 581ms
_ YN0000: Failed with errors in 0s 583ms
```

### To reproduce

In any yarn project:

```
$ yarn add flexsearch@skiff-org/flexsearch
```

### Environment

System:
  - MacOS 11.5.1
Binaries:
  - node v16.6.1
  - yarn 1.22.10

**Comment by arcanis** [comment]
This has been fixed in 3.0 (https://github.com/yarnpkg/berry/commit/6bdf043bfee35ef61df2121885b3a086bb77154b).

> yarn 1.22.10

This isn't the right version.

**Comment by ndrewtl** [comment]
Yes, my mistake, I'm on 2.4.1. Thanks!

**Comment by dcastil** [comment]
I'm still experiencing this issue on yarn v3.0.2. Is there a way for me to fix this issue or will this only be included in a future release?

My current workaround is to temporarily downgrade to yarn v1 when I upgrade my package version, but that's quite cumbersome. 😅

**Comment by arcanis** [comment]
As mentioned, it's been fixed in 3.0. Double check your version and open a new issue (with a repro) if necessary.
